src/Microsoft.AspNetCore.Mvc.ApiExplorer/ApiResponseTypeProvider.cs

```diff
@@ -0,0 +1,328 @@
+                            contentType,
+                            objectType.Value);
+
+                        if (formatterSupportedContentTypes == null)
+                        {
+                            continue;
+                        }
+
+                        foreach (var formatterSupportedContentType in formatterSupportedContentTypes)
+                        {
+                            apiResponseType.ApiResponseFormats.Add(new ApiResponseFormat()
+                            {
+                                Formatter = (IOutputFormatter)responseTypeMetadataProvider,
+                                MediaType = formatterSupportedContentType,
+                            });
+                        }
+                    }
+                }
+
+                results.Add(apiResponseType);
+            }
+
+            return results;
+        }
+
+        private Type GetDeclaredReturnType(ControllerActionDescriptor action)
+        {
+            var declaredReturnType = action.MethodInfo.ReturnType;
+            if (declaredReturnType == typeof(void) ||
+                declaredReturnType == typeof(Task))
+            {
+                return typeof(void);
+            }
+
+            // Unwrap the type if it's a Task<T>. The Task (non-generic) case was already handled.
+            Type unwrappedType = declaredReturnType;
+            if (declaredReturnType.IsGenericType &&
+                declaredReturnType.GetGenericTypeDefinition() == typeof(Task<>))
+            {
+                unwrappedType = declaredReturnType.GetGenericArguments()[0];
+            }
+
+            // If the method is declared to return IActionResult or a derived class, that information
+            // isn't valuable to the formatter.
+            if (typeof(IActionResult).IsAssignableFrom(unwrappedType))
+            {
+                return null;
+            }
+
+            // If we get here, the type should be a user-defined data type or an envelope type
+            // like ActionResult<T>. The mapper service will unwrap envelopes.
+            unwrappedType = _mapper.GetResultDataType(unwrappedType);
+            return unwrappedType;
+        }
+
+        private Type GetRuntimeReturnType(Type declaredReturnType)
+        {
+            // If we get here, then a filter didn't give us an answer, so we need to figure out if we
+            // want to use the declared return type.
+            //
+            // We've already excluded Task, void, and IActionResult at this point.
+            //
+            // If the action might return any object, then assume we don't know anything about it.
+            if (declaredReturnType == typeof(object))
+            {
+                return null;
+            }
+
+            return declaredReturnType;
+        }
+
+        internal static bool IsMatch(MethodInfo methodInfo, MethodInfo conventionMethod)
```

Comment:
Should we extract the code from here on to a separate abstraction to allow users to tweak?

Reply:
In general, we want the runtime behavior to be tightly coupled with the way analyzers work. The moment the two implementations diverge, you no longer get correct diagnosis which would make the tooling experience frustrating.

Reply:
This would become a common question going forward I think. We provide a default behavior based on which we provide analyzers and sometimes users want to change the default behavior and what happens to analyzers in that case. Maybe users have to disable them explicitly for certain parts/components. how easy is it to do it, i do not know. I am not suggesting you do changes here.

Reply:
I agree with what @kichalla - except for the abstraction part. This could be its own thing that builds an object model - or it could be part of application model (if it's reasonable).

Reply:
Ok. I'll get this change in and follow up with moving this code in to to the application model. The plan is for users to live with the consequences if their custom convention behavior no longer matches the tooling.

Reply:
Filed https://github.com/aspnet/Mvc/issues/7906. We can consider this after preview1 once we've had feedback.